### PR TITLE
fix(auth): remove the ability to have auth chained with each request

### PR DESCRIPTION
**Pull request by domharrington**
## 🧰 Changes
There's [a bug](https://github.com/readmeio/api/issues/102) with chaining auth calls due to the fact that we have a global state, and calling auth() again appends on to the state and breaks prepareAuth.

This PR removes the ability to call auth() in a chainable fashion, and only allows auth to be called at the top level.

I'd like to figure out a way to bring this back eventually, but the current API design doesn't support it in a clean way. We can revisit this when we look to support operations with multiple auth mechanisms.

This should probably be a major version bump due to removing functionality.

Fixes https://github.com/readmeio/api/issues/102

## 🧬 QA & Testing

Does CI pass? 🟢

### CODE REVIEW

File: packages/api/src/core/index.ts [hunk 1]
```diff
@@ -26,7 +26,7 @@ export { getJSONSchemaDefaults, parseResponse, prepareAuth, prepareParams, prepa
 export default class APICore {
   spec: Oas;
 
-  private auth: (number | string)[][] = [];
+  private auth: (number | string)[] = [];
```
Comment (domharrington) [inline packages/api/src/core/index.ts:29]: Did i typescript correctly?

File: packages/api/src/core/index.ts [hunk 2]
```diff
@@ -59,7 +59,7 @@ export default class APICore {
   }
 
   setAuth(...values: string[] | number[]) {
-    this.auth.push(values);
+    this.auth = values;
     return this;
```
Comment (domharrington) [inline packages/api/src/core/index.ts:63]: Not sure why these are all returning `this`. I don't think it needs to, but I kept it to be consistent with the others.

File: packages/api/src/core/prepareAuth.ts [hunk 3]
```diff
@@ -27,60 +27,59 @@ export default function prepareAuth(authKeys: (number | string)[][], operation:
...
-        break;
+  const securityType = securitySchemes[0] as SecurityType;
+
+  const schemes = security[securityType];
+
+  if (schemes.length > 1) {
+    throw new Error("Sorry, this API currently requires multiple forms of authentication which we don't yet support.");
+  }
+
+  const scheme = schemes[0];
```
Comment (domharrington) [inline packages/api/src/core/prepareAuth.ts:38]: Same as https://github.com/readmeio/api/pull/431/files#r847494000=

File: packages/api/src/index.ts [hunk 4]
```diff
@@ -156,7 +156,6 @@ class Sdk {
        */
       auth: (...values: string[] | number[]) => {
         core.setAuth(...values);
-        return new Proxy(sdk, sdkProxy);
```
Comment (domharrington) [inline packages/api/src/index.ts:159]: This is the part that actually removes the chaining.

**Review by domharrington — COMMENTED** [review]
- [ ] Can I remove this part here as well? https://github.com/readmeio/api/blob/15bfc55bfecdff4d336fa632a2e6370008e77d3b/packages/api/src/index.ts#L104
- [ ] I've left the tests skipped, up to you if you wanna keep em around or if we should just remove them

File: packages/api/src/core/index.ts (hunk 1 above)
Comment (erunion) [reply packages/api/src/core/index.ts:29]: Yep. I updated `prepareAuth.test.ts` in https://github.com/readmeio/api/pull/431/commits/091049816549923dc1069df2ea6d679fa4d075f1 to account for this.

File: packages/api/src/core/index.ts (hunk 2 above)
Comment (erunion) [reply packages/api/src/core/index.ts:63]: I think this was my OOP brain trying to make things chainable and I don't think any of these `set*(` methods need to be.

File: packages/api/src/core/prepareAuth.ts [hunk 5]
```diff
@@ -27,60 +27,59 @@ export default function prepareAuth(authKeys: (number | string)[][], operation:
...
-          if (authKey.length > 1) {
-            throw new Error(
-              'Multiple auth tokens were supplied for the auth on this endpoint, but only a single token is needed.'
-            );
-          }
-
-          preparedAuth[scheme._key] = authKey[0];
-        }
-        break;
+  const securityType = securitySchemes[0] as SecurityType;
```
[the remark this answers is not in the document]
Comment (erunion) [reply packages/api/src/core/prepareAuth.ts:30]: Yeah this is fine.

File: packages/api/src/index.ts (hunk 4 above)
Comment (erunion) [reply packages/api/src/index.ts:159]: Can you update the README? There's a handful of references in there talking about how `.auth()` is chainable. You'll need to update the root README and the one in `packages/api` too but I believe the root one is just a copy of the package one.

File: packages/api/test/auth.test.ts [hunk 6]
```diff
@@ -29,30 +29,29 @@ describe('#auth()', function () {
...
+        const mock = nock('https://httpbin.org')
+          .get(/\/anything\/apiKey/)
+          .reply(200, uri => uri);
+
+        sdk.auth(apiKey);
+        expect(await sdk.get('/anything/apiKey')).to.equal('/anything/apiKey?apiKey=123457890');
+        mock.done();
+      });
+
+      it.skip('should allow you to supply auth when chained with an operation', async function () {
```
Comment (erunion) [inline packages/api/test/auth.test.ts:42]: I'm good with removing all these skipped tests.

**Review by erunion — CHANGES_REQUESTED** [review]
need to update the docs on `.auth()` but otherwise this looks good.

**Comment by TomMiller-mas** [comment]
Late to the party, but what about letting us close the connection manually when we are done with the SDK connection?  

`await sdk.close();`

This allows us to leave it open as long as we want to.

**Pull request merged**